File: backend/scripts/update_models.py
import json
import urllib.request
import urllib.error

# 读取 db.json
with open('backend/data/db.json', 'r', encoding='utf-8') as f:
    db = json.load(f)

# 从远程获取模型列表
url = 'http://localhost:8080/v1/models'
req = urllib.request.Request(url, headers={
    'Authorization': 'Bearer 13391822168',
    'Accept': 'application/json',
    'User-Agent': 'OpenClaw-Backend-Agent'
})

try:
    with urllib.request.urlopen(req) as response:
        data = json.loads(response.read().decode('utf-8'))
        remote_models = data.get('data', data)
        print(f"[Fetch] 获取到 {len(remote_models)} 个远程模型")
except urllib.error.HTTPError as e:
    print(f"[Error] HTTP {e.code}: {e.read().decode('utf-8')}")
    exit(1)

# 保留现有模型的温度/maxTokens设置
existing_models = db.get('availableModels', [])
existing_map = {m['modelId']: m for m in existing_models}

# 过滤保留非 glue provider 的模型
non_glue_models = [m for m in existing_models 
                    if m.get('provider') != 'glue' and 'localhost:8080' not in (m.get('baseUrl') or '')]

# 构建新模型列表 (glue provider)
new_models = []
for m in remote_models:
    model_id = m.get('id') or m.get('modelId') or ''
    existing = existing_map.get(model_id, {})
    
    new_models.append({
        'id': model_id,
        'name': m.get('name') or model_id,
        'modelId': model_id,
        'baseUrl': 'http://localhost:8080/v1',
        'apiKey': '13391822168',
        'provider': 'glue',
        'temperature': existing.get('temperature', 0.7),
        'maxTokens': existing.get('maxTokens', 4096),
        'description': m.get('description') or f"Via glue proxy: {model_id}"
    })

# 合并模型列表
db['availableModels'] = non_glue_models + new_models

# 保存
with open('backend/data/db.json', 'w', encoding='utf-8') as f:
    json.dump(db, f, ensure_ascii=False, indent=2)

print(f"[Success] 已更新模型列表")
print(f"  - 保留非 glue 模型: {len(non_glue_models)}")
print(f"  - 新增 glue 模型: {len(new_models)}")
print(f"  - 总计: {len(db['availableModels'])}")

File: backend/scripts/update-models.ts
import { readFile, writeFile } from 'node:fs/promises';
import * as path from 'node:path';

const DB_PATH = path.resolve(process.cwd(), 'data/db.json');

async function fetchRemoteModels() {
  const baseUrl = 'http://localhost:8080/v1';
  const url = `${baseUrl}/models`;
  const apiKey = '13391822168';

  console.log(`[Fetch] 正在从 ${url} 获取模型列表...`);

  const response = await fetch(url, {
    method: 'GET',
    headers: {
      'Authorization': `Bearer ${apiKey}`,
      'Accept': 'application/json',
      'User-Agent': 'OpenClaw-Backend-Agent'
    }
  });

  if (!response.ok) {
    const status = response.status;
    const errorText = await response.text();
    console.error(`[Fetch] 远程服务报错 (${status}): ${errorText}`);
    throw new Error(`HTTP ${status}: ${errorText}`);
  }

  const data: any = await response.json();
  const models = data.data || data;

  if (!Array.isArray(models)) {
    throw new Error('返回数据格式不符合模型列表规范');
  }

  console.log(`[Fetch] 获取到 ${models.length} 个模型`);
  return models;
}

async function updateModels() {
  // 加载数据库
  const dbContent = await readFile(DB_PATH, 'utf-8');
  const db = JSON.parse(dbContent);

  // 获取远程模型列表
  const remoteModels = await fetchRemoteModels();

  // 保留现有模型的温度/maxTokens设置
  const existingModels = db.availableModels || [];
  const existingMap = new Map(existingModels.map(m => [m.modelId, m]));

  // 过滤出 glue provider 的模型 (baseUrl 包含 localhost:8080)
  const glueModels = remoteModels.filter(m => {
    const id = m.id || m.modelId || '';
    // 远程API返回的模型ID格式可能是 "provider/model-id"
    return !id.includes('/') || id.startsWith('glue/') || id.startsWith('claude-') || id.startsWith('gpt-') || id.startsWith('gemini-') || id.startsWith('deepseek-') || id.startsWith('minimax');
  });

  console.log(`[Filter] 符合 glue provider 的模型: ${glueModels.length} 个`);

  // 构建新模型列表
  const newModels = remoteModels.map(m => {
    const modelId = m.id || m.modelId || '';
    const existing = existingMap.get(modelId);
    
    return {
      id: modelId,
      name: m.name || modelId,
      modelId: modelId,
      baseUrl: 'http://localhost:8080/v1',
      apiKey: '13391822168',
      provider: 'glue',
      temperature: existing?.temperature ?? 0.7,
      maxTokens: existing?.maxTokens ?? 4096,
      description: m.description || `Via glue proxy: ${modelId}`
    };
  });

  console.log(`[Update] 更新模型数量: ${db.availableModels?.length || 0} -> ${newModels.length}`);

  // 保留非 glue provider 的模型
  const nonGlueModels = (db.availableModels || []).filter(m => 
    m.provider !== 'glue' && !m.baseUrl?.includes('localhost:8080')
  );

  db.availableModels = [...nonGlueModels, ...newModels];

  // 保存
  await writeFile(DB_PATH, JSON.stringify(db, null, 2), 'utf-8');
  console.log(`[Success] 已更新模型列表，共 ${db.availableModels.length} 个模型`);

  return db.availableModels;
}

updateModels().catch(err => {
  console.error('[Error]', err.message);
  process.exit(1);
});